app/feature/feature-movingflow/src/main/kotlin/com/hedvig/android/feature/movingflow/ui/selectcontract/SelectContractViewModel.kt

```diff
@@ -0,0 +1,189 @@
+  )
+
+internal class SelectContractPresenter(
+  private val apolloClient: ApolloClient,
+  private val movingFlowRepository: MovingFlowRepository,
+) : MoleculePresenter<SelectContractEvent, SelectContractState> {
+  @Composable
+  override fun MoleculePresenterScope<SelectContractEvent>.present(
+    lastState: SelectContractState,
+  ): SelectContractState {
+    var loadIteration by remember { mutableIntStateOf(0) }
+    var currentState by remember { mutableStateOf(lastState) }
+    var submittingAddressId: String? by remember { mutableStateOf(null) }
+
+    CollectEvents { event ->
+      when (event) {
+        SelectContractEvent.ClearNavigation -> {
+          val state = currentState as? Content ?: return@CollectEvents
+          currentState = state.copy(navigateToHousingType = false)
+        }
+
+        is SelectContractEvent.SelectContract -> {
+          val state = currentState as? Content ?: return@CollectEvents
+          currentState = state.copy(
+            selectedAddress = event.contract,
+          )
+        }
+
+        SelectContractEvent.SubmitContract -> {
+          val state = currentState as? Content ?: return@CollectEvents
+          val selectedId = state.selectedAddress ?: return@CollectEvents
+          submittingAddressId = selectedId.id
+        }
+
+        SelectContractEvent.RetryLoadData -> loadIteration++
+      }
+    }
+
+    LaunchedEffect(submittingAddressId) {
+      val id = submittingAddressId
+      if (id != null) {
+        val state = currentState as? NotEmpty ?: return@LaunchedEffect
+        if (state is Content) {
+          currentState = state.copy(buttonLoading = true)
+        }
+        val moveIntent = state.intent
+        movingFlowRepository.initiateNewMovingFlow(moveIntent, id)
+        submittingAddressId = null
+        currentState = when (state) {
+          is Content -> state.copy(navigateToHousingType = true, buttonLoading = false)
+          is Redirecting -> state.copy(navigateToHousingType = true)
+        }
+      }
+    }
+
+    LaunchedEffect(loadIteration) {
+      either {
+        val moveIntentCreate = apolloClient
+          .mutation(MoveIntentV2CreateMutation())
+          .safeExecute()
+          .mapLeft(::ErrorMessage)
+          .mapLeft { SelectContractState.Error.GenericError(it) }
+          .map { it.moveIntentCreate }
+          .bind()
+        val moveIntent = ensureNotNull(moveIntentCreate.moveIntent) {
+          val userError = moveIntentCreate.userError?.message
+          if (userError == null) {
+            SelectContractState.Error.GenericError(
+              ErrorMessage(
+                "Unknown MoveIntentV2CreateMutation error",
+              ),
+            )
+          } else {
+            SelectContractState.Error.UserPresentable(userError)
+          }
+        }
+        moveIntent
+      }.fold(
+        ifLeft = { error ->
+          Snapshot.withMutableSnapshot {
+            currentState = error
+          }
```

Comment:
Since we only edit 1 mutableState here, we do not risk there being inconsistent states where one thing is edited before the other thing also has a chance to be edited.
```suggestion
          currentState = error
```